apache/metron
# METRON-2065: Setting Parser Output Topic in Sensor Config is broken

## Contributor Comments
This PR fixes a minor bug that fails to honor the `outputTopic` parser configuration setting and adds a unit test to cover this method.  I slightly refactored the `ParserTopologyBuilder` class to make this feature easier to test.

### Testing
This has been tested in full dev.

1. Spin up full dev and ensure bro, snort and yaf data is landing in the `enrichments` topic with the Kafka console consumer tool:
```
/usr/hdp/current/kafka-broker/bin/kafka-console-consumer.sh --zookeeper node1:2181 --topic enrichments
```
2. Update the bro parser configuration to write to a custom output topic:
```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{
  "parserClassName": "org.apache.metron.parsers.bro.BasicBroParser",
  "sensorTopic": "bro",
  "outputTopic": "bro_test"
}' 'http://user:password@node1:8082/api/v1/sensor/parser/config/bro'
```
3. Update the snort parser configuration to write to a custom output topic:
```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{
  "parserClassName": "org.apache.metron.parsers.snort.BasicSnortParser",
  "sensorTopic": "snort",
  "outputTopic": "snort_test"
}' 'http://user:password@node1:8082/api/v1/sensor/parser/config/snort'
```
4. Update the yaf parser configuration to write to a custom output topic:
```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{
  "parserClassName": "org.apache.metron.parsers.GrokParser",
  "sensorTopic": "yaf",
  "outputTopic": "yaf_test",
  "parserConfig": {
    "grokPath": "/patterns/yaf",
    "patternLabel": "YAF_DELIMITED",
    "timestampField": "start_time",
    "timeFields": [
      "start_time",
      "end_time"
    ],
    "dateFormat": "yyyy-MM-dd HH:mm:ss.S"
  },
  "fieldTransformations": [
    {
      "input": [
        "protocol"
      ],
      "output": [
        "protocol"
      ],
      "transformation": "IP_PROTOCOL",
      "config": {}
    }
  ]
}' 'http://user:password@node1:8082/api/v1/sensor/parser/config/yaf'
```
5. Restart the parser topology component in Ambari.
6. After the topology has started back up, there should be 3 new `*_test` topics:
```
[root@node1 vagrant]# /usr/hdp/current/kafka-broker/bin/kafka-topics.sh --zookeeper node1:2181 --list
__consumer_offsets
bro
bro_test
enrichments
escalation
indexing
pcap
snort
snort_test
yaf
yaf_test
```
7. These new topics should all have data written to them while the `enrichments` topic should no longer be written to.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:


### For code changes:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```


### For documentation related changes:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.
